File: backend/api/routes/retrieval_lab.py
from fastapi import APIRouter

from schemas.retrieval_lab import RetrievalCompareRequest, RetrievalCompareResponse
from services.retrieval_lab_service import compare_retrieval_modes

router = APIRouter(prefix="/repositories/{repository_id}/retrieval", tags=["retrieval-lab"])


@router.post("/compare", response_model=RetrievalCompareResponse)
def compare_repository_retrieval(
    repository_id: int,
    payload: RetrievalCompareRequest,
) -> RetrievalCompareResponse:
    """Compare retrieval strategies for the same question (no LLM call)."""
    result = compare_retrieval_modes(
        repository_id,
        message=payload.message,
        top_k=payload.top_k,
    )
    return RetrievalCompareResponse.model_validate(result)

File: frontend/src/api/retrievalLab.ts
import type { RetrievalCompareRequest, RetrievalCompareResponse } from '@/types/retrievalLab'
import { apiClient, getApiErrorMessage } from '@/lib/axios'

export async function compareRetrievalModes(
  repositoryId: number,
  payload: RetrievalCompareRequest,
) {
  try {
    const { data } = await apiClient.post<RetrievalCompareResponse>(
      `/api/v1/repositories/${repositoryId}/retrieval/compare`,
      payload,
    )
    return data
  } catch (error) {
    throw new Error(getApiErrorMessage(error, 'Retrieval comparison failed'))
  }
}
